From e4ca5abb8e795623310364318100f9e5501aca29 Mon Sep 17 00:00:00 2001
From: Guruswamy Basavaiah <guruswamy.basavaiah@broadcom.com>
Date: Wed, 21 May 2025 08:52:48 +0000
Subject: [PATCH] kpatch: compatibility with Photon gcc RAP patch

In commit [1], kpatch added support for function padding,
and CONFIG_CFI_CLANG, which hardcoded a value of 16 for
the prefix size.

The length of padding around __cfi prefixed functions can vary,
depending on certain things such as config options, like
CONFIG_CALL_PADDING [2]. In Photon, with our patch
SPECS/linux/secure/gcc-rap-plugin-with-kcfi.patch,
we also add extra int3 (0xcc) instructions to the __cfi
function padding, but without the NOPs.

I don't see any good reason to keep this as hardcoded,
so let's make it more flexible and use the prefix size
that we read in.

Also, we need to account for __rap_hash_##name format
objects appearing in the symbol table. kpatch will fail
as it tries to bundle these, but they have no section
info. They should not be bundled, nor are they dynamic
debug symbols, so patch accordingly to get kpatch to work.

Ref:
1. https://github.com/dynup/kpatch/commit/3e54c63b175b68cf48654c119e62bda398d0c018
2. https://github.com/torvalds/linux/blob/master/arch/x86/include/asm/cfi.h#L51

Signed-off-by: Brennan Lamoreaux <brennan.lamoreaux@broadcom.com>
[Guru: Ported this patch to kpatch v0.9.10. Removed changes
from patch which are present in kpatch version v0.9.10
Signed-off-by: Guruswamy Basavaiah <guruswamy.basavaiah@broadcom.com>
---
 kpatch-build/create-diff-object.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/kpatch-build/create-diff-object.c b/kpatch-build/create-diff-object.c
index 8b567f4..20091e2 100644
--- a/kpatch-build/create-diff-object.c
+++ b/kpatch-build/create-diff-object.c
@@ -91,6 +91,21 @@ struct special_section {
 
 static bool is_bundleable(struct symbol *sym)
 {
+	/*
+	 * Photon patch for RAP plugin will link
+	 * global objects with the prefix __rap_hash_##func_name
+	 * into the symbol table.
+	 *
+	 * I believe these are the hashed RAP values,
+	 * so there is no need to bundle these. They are
+	 * not bundleable since they have no section info.
+	 *
+	 * Besides, in any other cases, if there is no section info
+	 * then we can't bundle it anyhow.
+	 */
+	if (sym->type == STT_OBJECT && !sym->sec)
+		return false;
+
 	if (sym->type == STT_FUNC &&
 	    !strncmp(sym->sec->name, ".text.",6) &&
 	    !strcmp(sym->sec->name + 6, sym->name))
@@ -315,6 +330,8 @@ static void kpatch_detect_child_functions(struct kpatch_elf *kelf)
 
 static bool is_dynamic_debug_symbol(struct symbol *sym)
 {
+	if (sym->type == STT_OBJECT && !strncmp(sym->name, "__rap_hash", 10))
+		return false;
 	if (sym->type == STT_OBJECT && !strcmp(sym->sec->name, "__verbose"))
 		return true;
 	if (sym->type == STT_OBJECT && !strcmp(sym->sec->name, "__dyndbg"))
-- 
2.40.4

